server/utils/serviceProviderCSVGenerator.ts

```diff
@@ -0,0 +1,54 @@
+import stringify from 'csv-stringify'
+import { ServiceProviderReportReferral } from '../models/serviceProviderReportReferral'
+
+export default class ServiceProviderCSVGenerator {
+  static generate(reportReferrals: ServiceProviderReportReferral[]): string {
+    const data: string[] = []
+    const stringifier = stringify({
+      header: true,
+      columns: [
+        { key: 'referralLink', header: 'referral_link' },
+        { key: 'referralRef', header: 'referral_ref' },
+        { key: 'referralId', header: 'referral_id' },
+        { key: 'contractId', header: 'contract_id' },
+        { key: 'organisationId', header: 'organisation_id' },
+        { key: 'referringOfficerEmail', header: 'referring_officer_email' },
+        { key: 'caseworkerId', header: 'caseworker_id' },
+        { key: 'serviceUserCRN', header: 'service_user_crn' },
+        { key: 'dateReferralReceived', header: 'date_referral_received' },
+        { key: 'dateSAABooked', header: 'date_saa_booked' },
+        { key: 'dateSAAAttended', header: 'date_saa_attended' },
+        { key: 'dateFirstActionPlanSubmitted', header: 'date_first_action_plan_submitted' },
+        { key: 'dateOfFirstActionPlanApproval', header: 'date_of_first_action_plan_approval' },
+        { key: 'dateOfFirstAttendedSession', header: 'date_of_first_attended_session' },
+        { key: 'outcomesToBeAchievedCount', header: 'outcomes_to_be_achieved_count' },
+        { key: 'outcomesAchieved', header: 'outcomes_achieved' },
+        { key: 'countOfSessionsExpected', header: 'count_of_sessions_expected' },
+        { key: 'countOfSessionsAttended', header: 'count_of_sessions_attended' },
+        { key: 'endRequestedByPPAt', header: 'end_requested_by_pp_at' },
+        { key: 'endRequestedByPPReason', header: 'end_requested_by_pp_reason' },
+        { key: 'dateEOSRSubmitted', header: 'date_eosr_submitted' },
+        { key: 'concludedAt', header: 'concluded_at' },
+      ],
+    })
+
+    stringifier.on('readable', () => {
+      let row = stringifier.read()
+      while (row) {
+        data.push(row)
+        row = stringifier.read()
+      }
+    })
+
+    reportReferrals.forEach(referral => {
+      stringifier.write(referral)
+    })
+    stringifier.end()
```

Comment:
I'm a little confused by how this method is working. As far as I can tell, this streaming API provided by `stringifier` is an _asynchronous_ API - you:

1. give it a bunch of data (`stringifier.write()`)
2. tell it you've given it all the data that you've got to offer (`stringifier.end()`)

and then, over time, it incrementally generates the CSV and calls you back with a handful of lines at a time (`stringifier.on('readable', …)`). It then eventually tells you that it's done via another callback (I think that's `stringifier.on('finish', …)`, seen in [the library's docs](https://csv.js.org/stringify/api/)), but I don't see us using that callback here.

So, if this whole process is asynchronous, then I think that the method which wraps this process also needs to be asynchronous. It feels like the signature for this method should be 

```typescript
static async generate(reportReferrals: ServiceProviderReportReferral[]): Promise<string>
```

and we need to return a promise which we resolve upon receiving the `finish` callback.

On the other hand, this does seem to be working, at least for a single row in the tests, but my feeling is that if we're not using the API as intended then it could stop working at any time or not work for larger data sets.

Reply:
I tried increasing the length of the array in the test to ~200 and it stops working - just outputs an empty string. I'll take a look into this, but presumably related to the above.